### PR TITLE
add migration stats when collecting statistics about the trie

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Reasoning behind the pull request
- In order to track the migration of trie nodes from one version to another, collect stats about trie nodes version.
  
## Proposed changes
- Collect stats about trie nodes version during state snapshot.

## Testing procedure
- Normal testing procedure

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: trie/statistics/trieStatistics.go [hunk 1]
```diff
@@ -101,12 +106,13 @@ type TrieStatsDTO struct {
...
-	NumExtensionNodes  uint64
-	LeafNodesSize      uint64
-	NumLeafNodes       uint64
+	BranchNodesSize      uint64
+	NumBranchNodes       uint64
+	ExtensionNodesSize   uint64
+	NumExtensionNodes    uint64
+	LeafNodesSize        uint64
+	NumLeafNodes         uint64
+	LeavesMigrationStats map[core.TrieNodeVersion]uint64
```
Comment (iulianpascalau) [inline trie/statistics/trieStatistics.go:115]: a little bit dangerous to use a map in a plain DTO.
Reason: this DTO can be passed among multiple structs. If we happen to do a write in the map while executing a read, a panic can occur. For the other fields, the code won't panic event if the concurrency issue happens. Suggestion to refactor the solution.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: trie/statistics/trieStatistics.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply trie/statistics/trieStatistics.go]: Refactored.

File: trie/statistics/trieStatistics.go [hunk 2]
```diff
@@ -73,55 +91,125 @@ func (ts *trieStatistics) collectNodeStatistics(level int, size uint64, nodeStat
...
-	RootHash       []byte
-	TotalNodesSize uint64
-	TotalNumNodes  uint64
-	MaxTrieDepth   uint32
+// GetLeavesMigrationStats will return the leaves migration statistics
+func (ts *trieStatistics) GetLeavesMigrationStats() map[core.TrieNodeVersion]uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.migrationStats
```
Comment (iulianpascalau) [inline trie/statistics/trieStatistics.go:179]: we need to return a new map here. Otherwise, the mutex protection is useless.

Comment (BeniaminDrasovean) [reply trie/statistics/trieStatistics.go]: Refactored.

**Review by iulianpascalau — DISMISSED** [review]

File: trie/statistics/trieStatisticsCollector.go [hunk 3]
```diff
@@ -38,35 +43,45 @@ func NewTrieStatisticsCollector() *trieStatisticsCollector {
...
+		triesBySize:         make([]common.TrieStatisticsHandler, numTriesToPrint),
+		triesByDepth:        make([]common.TrieStatisticsHandler, numTriesToPrint),
+		migrationStats:      make(map[core.TrieNodeVersion]uint64),
 	}
 }
 
 // Add adds the given trie statistics to the statistics collector
-func (tsc *trieStatisticsCollector) Add(trieStats *TrieStatsDTO) {
+func (tsc *trieStatisticsCollector) Add(trieStats common.TrieStatisticsHandler) {
 	if trieStats == nil {
```
Comment (SebastianMarian) [inline trie/statistics/trieStatisticsCollector.go:54]: check.IfNil ?

File: trie/statistics/trieStatisticsCollector.go [hunk 4]
```diff
@@ -80,17 +95,21 @@ func (tsc *trieStatisticsCollector) Print() {
...
+	defer tsc.mutex.RUnlock()
+	
 	return tsc.numNodes
 }
 
-func getOrderedTries(tries []*TrieStatsDTO) string {
+func getOrderedTries(tries []common.TrieStatisticsHandler) string {
 	triesStats := make([]string, 0)
 	for i := 0; i < len(tries); i++ {
 		if tries[i] == nil {
```
Comment (SebastianMarian) [inline trie/statistics/trieStatisticsCollector.go:115]: check.IfNil ?

File: trie/statistics/trieStatisticsCollector.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply trie/statistics/trieStatisticsCollector.go:54]: Done.

File: trie/statistics/trieStatisticsCollector.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply trie/statistics/trieStatisticsCollector.go:115]: Done.

**Review by SebastianMarian — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**